rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -21,92 +21,152 @@
 import co.rsk.util.MaxSizeHashMap;
 import org.ethereum.db.ByteArrayWrapper;
 import org.ethereum.util.ByteUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.*;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 public class DataSourceWithCache implements KeyValueDataSource {
+    private static final Logger logger = LoggerFactory.getLogger(DataSourceWithCache.class);
+
+    private final int cacheSize;
     private final KeyValueDataSource base;
     private final Map<ByteArrayWrapper, byte[]> uncommittedCache;
     private final Map<ByteArrayWrapper, byte[]> committedCache;
 
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
     public DataSourceWithCache(KeyValueDataSource base, int cacheSize) {
+        this.cacheSize = cacheSize;
         this.base = base;
-        this.uncommittedCache = new HashMap<>();
-        this.committedCache = new MaxSizeHashMap<>(cacheSize, true);
+        this.uncommittedCache = new LinkedHashMap<>(cacheSize / 8, (float)0.75, false);
+        this.committedCache = Collections.synchronizedMap(new MaxSizeHashMap<>(cacheSize, true));
```

Comment:
Having a synchronized map here will needlessly lock every get to the `comittedCache` and make the multiple concurrency of `DataSourceWithCache#get()` useless.

I'd rather use the `WriteLock`, a `ConcurrentMap` or another `Lock`, the problem with `concurrentMap` is that we are losing the `MaxSize` property. 
We could roll our own `concurrentMasSizeHashMap` but that might be out of the scope of this PR, so i'd just use the `WriteLock`.

Reply:
Add write lock, done!